File: lib/core_kernel/token_tree.ml
(* The reader's output: tokens and delimiter groups, each token carrying its
   scope set. In the kernel so unparsed bodies can sit inside [Syntax.t]. *)

type delimiter = Paren | Bracket | Brace [@@deriving show]

type token_kind =
  | Ident of string
  | Int of int64
  | Char of char
  | String of string
  | Unit
  | KwLet
  | KwFun
  | KwThen
  | KwSig
  | KwFn
  | KwDo
  | KwEnd
  | KwElse
  | KwMatch
  | KwWith
  | KwEffect
  | KwModule
  | KwStruct
  | KwEnum
  | KwImpl
  | KwTrait
  | KwPub
  | KwImport
  | KwOpen
  | KwMacro
  | KwPattern
  | KwSelf
  | KwSelfType
  | KwRef
  | KwDeref
  | KwRec
  | KwCan
  | KwPerform
  | KwResume
  | KwMethod
  | KwUnit
  | LParen
  | RParen
  | LBracket
  | RBracket
  | LBrace
  | RBrace
  | Comma
  | Dot
  | Colon
  | Equals
  | Semi
  | Bar
  | ThinArrow
  | DatumComment
  | Operator of string
  | Eof
[@@deriving show]

(* [scope] is the token's scope set: the reader gives every token the empty
   set; enforestation adds the scopes of the definition contexts, template
   instances and binders around it before the token is read as a form, so a
   syntactic role is resolved against the token's scopes (M7). *)
type token = {
  kind : token_kind;
  span : Source_span.t;
  scope : Scope_set.t; [@opaque]
}
[@@deriving show]

type datum =
  | Token of token
  | Group of delimiter * t list * Source_span.t
[@@deriving show]

and t = {
  datum : datum;
  span : Source_span.t;
}
[@@deriving show]

let token ?(scope = Scope_set.empty) kind span = { kind; span; scope }

let syntax_token ?scope kind span = { datum = Token (token ?scope kind span); span }

(* Add [s] to the scope set of every token in [terms], inside groups too. *)
let rec add_scope (s : Scope_set.t) (terms : t list) : t list =
  List.map
    (fun term ->
      match term.datum with
      | Token tok -> { term with datum = Token { tok with scope = Scope_set.union tok.scope s } }
      | Group (d, items, span) -> { term with datum = Group (d, add_scope s items, span) })
    terms

let group delimiter items span = { datum = Group (delimiter, items, span); span }


(* Every keyword and punctuation token by its spelling, both ways: reflection
   carries these tokens as their spelling (M9). *)
let keyword_spellings =
  [ (KwLet, "let"); (KwFun, "fun"); (KwThen, "then"); (KwSig, "sig"); (KwFn, "fn"); (KwDo, "do");
    (KwEnd, "end"); (KwElse, "else"); (KwMatch, "match"); (KwWith, "with"); (KwEffect, "effect");
    (KwModule, "module"); (KwStruct, "struct"); (KwEnum, "enum"); (KwImpl, "impl"); (KwTrait, "trait");
    (KwPub, "pub"); (KwImport, "import"); (KwOpen, "open"); (KwMacro, "macro"); (KwPattern, "pattern");
    (KwSelf, "self"); (KwSelfType, "Self"); (KwRef, "ref"); (KwDeref, "deref"); (KwRec, "rec");
    (KwCan, "can"); (KwPerform, "perform"); (KwResume, "resume"); (KwMethod, "method"); (KwUnit, "Unit") ]

let punct_spellings =
  [ (LParen, "("); (RParen, ")"); (LBracket, "["); (RBracket, "]"); (LBrace, "{"); (RBrace, "}");
    (Comma, ","); (Dot, "."); (Colon, ":"); (Equals, "="); (Semi, ";"); (Bar, "|"); (ThinArrow, "->");
    (DatumComment, "#_"); (Eof, "EOF") ]

let spelling_of table kind = List.assoc_opt kind table

let of_spelling table s = Option.map fst (List.find_opt (fun (_, t) -> String.equal t s) table)

File: dotnet/src/Fun.Kernel/TokenTree.cs
using System.Collections.Immutable;

namespace Fun.Kernel;

public enum Delimiter { Paren, Bracket, Brace }

/// <summary>
/// A raw token. Keyword and punctuation kinds are singletons; the ones that
/// carry text or a literal are records.
/// </summary>
public abstract record TokenKind
{
    public sealed record Ident(string Name) : TokenKind;
    public sealed record Int(long Value) : TokenKind;
    public sealed record Char(char Value) : TokenKind;
    public sealed record Str(string Value) : TokenKind;
    public sealed record Operator(string Spelling) : TokenKind;

    /// <summary>A keyword or punctuation token, identified by its spelling.</summary>
    public sealed record Word(string Spelling) : TokenKind;

    // Keywords. `then`, `with`, `end`, `else` and `Unit` are deliberately absent:
    // they are ordinary identifiers (STATUS.md, stage 11 increment 2).
    public static readonly Word Let = new("let"), Fun = new("fun"), Sig = new("sig"),
        Fn = new("fn"), Do = new("do"), Match = new("match"), Effect = new("effect"),
        Module = new("module"), Struct = new("struct"), Enum = new("enum"),
        Impl = new("impl"), Trait = new("trait"), Pub = new("pub"), Import = new("import"),
        Open = new("open"), Export = new("export"), Macro = new("macro"),
        Pattern = new("pattern"), Self = new("self"), SelfType = new("Self"),
        Ref = new("ref"), Deref = new("deref"), Rec = new("rec"),
        Perform = new("perform"), Resume = new("resume"), Method = new("method");

    // Punctuation.
    public static readonly Word LParen = new("("), RParen = new(")"),
        LBracket = new("["), RBracket = new("]"), LBrace = new("{"), RBrace = new("}"),
        Comma = new(","), Dot = new("."), Colon = new(":"), Eq = new("="),
        Semi = new(";"), Bar = new("|"), ThinArrow = new("->"),
        DatumComment = new("#_"), Eof = new("EOF");

    /// <summary>The keywords, by spelling. Punctuation is matched by the reader directly.</summary>
    public static readonly ImmutableDictionary<string, Word> Keywords =
        new[] { Let, Fun, Sig, Fn, Do, Match, Effect, Module, Struct, Enum, Impl, Trait,
                Pub, Import, Open, Export, Macro, Pattern, Self, SelfType, Ref, Deref,
                Rec, Perform, Resume, Method }
            .ToImmutableDictionary(w => w.Spelling);

    /// <summary>What this token looks like in source, for diagnostics.</summary>
    public string Text() => this switch
    {
        Word w => w.Spelling,
        Ident i => i.Name,
        Operator o => o.Spelling,
        Int => "integer",
        Char => "char",
        Str => "string",
        _ => throw new InvalidOperationException($"unhandled token {GetType().Name}"),
    };
}

/// <summary>
/// A token with its scope set. The reader gives every token the empty set;
/// enforestation adds the scopes around it before it is read as a form, so a
/// syntactic role resolves against the token's scopes (M7).
/// </summary>
public sealed record Token(TokenKind Kind, SourceSpan Span, ScopeSet Scope)
{
    public Token(TokenKind kind, SourceSpan span) : this(kind, span, ScopeSet.Empty) { }
}

/// <summary>
/// The reader's output: tokens and delimiter groups. In the kernel so unparsed
/// bodies can sit inside a <see cref="Syntax"/> node.
/// </summary>
public abstract record TokenTree(SourceSpan Span)
{
    public sealed record Leaf(Token Token) : TokenTree(Token.Span);

    public sealed record Group(Delimiter Delimiter, ImmutableArray<TokenTree> Items, SourceSpan GroupSpan)
        : TokenTree(GroupSpan);

    /// <summary>Adds <paramref name="scope"/> to every token in the tree, inside groups too.</summary>
    public TokenTree AddScope(ScopeSet scope) => this switch
    {
        Leaf l => l with { Token = l.Token with { Scope = l.Token.Scope.Union(scope) } },
        Group g => g with { Items = [.. g.Items.Select(i => i.AddScope(scope))] },
        _ => throw new InvalidOperationException($"unhandled token tree {GetType().Name}"),
    };
}
